### ISSUE TITLE
Run Error using code you offer

### ISSUE THREAD

**Issue by JaysonChao** [LennonLab/MicrobialBiodiversityTheory#121]
Thanks for your rewriting the lognormal and other distribution patterns based on macroecological distribution module. There are some mistakes in the modelsAndMetrics script: no function called negbin_solver in macroecological distribution package but nbinom_lower_trunc_solver function, I think you mean this one when import the module. And also, the lognormal class ( lognorm_pred = mo.lognorm(SAD, 'pln')
lognorm_SAD = np.log10(lognorm_pred.get_rad_from_obs()[0]) ) used to creat figure1 is always wrong: /usr/local/lib/python2.7/site-packages/macroecotools-0.4.0-py2.7.egg/macroeco_distributions/macroeco_distributions.py:90: RuntimeWarning: overflow encountered in exp
You can try to run figure1 seprately and I think you will get the same error. Hoping you are doing well.

**Comment by wrshoemaker** [comment]
Hi Jayson,

Thank you for reporting the error. I looked into it and I think the problem is that the developers of the `macroeco_distributions` package have made a number of upgrades to the package since we used it. When I wrote the README for the package I thought I was careful enough to list the versions of all Python packages used in the study. Unfortunately I did not do that for `macroeco_distributions` or `macroecotools` and have since completely re-installed the operating system on my computer, so I can't just go into Python and type `import macroeco_distributions` to find out what version I used.

But the error `/usr/local/lib/python2.7/site-packages/macroecotools-0.4.0-py2.7.egg/macroeco_distributions/macroeco_distributions.py:90: RuntimeWarning: overflow encountered in exp` is usually a `SciPy` error that gets thrown up when a `SciPy` function is handling a number that is too large for an array. I didn't see this before and it might be a result of the changes made to `macroeco_distributions` over the last couple of years.

That said, what is your goal regarding the use of this code? Is it just to re-run our entire analysis, or are you only interested in fitting the Poisson lognormal to your data? I'm happy to help either way, I just want a clear picture since I think this is an issue of incompatibility between the newer versions of the packages we used and our code (and some poor documentation from my early years as a grad student) rather than mis-written code. If all you want to do is use the Poisson lognormal, that's an easier task than going back and re-writing and testing code for an already completed project.

Best,
Will

**Comment by JaysonChao** [comment]
Hi, Will,
Thanks for your help. I have successfuly generated figure 1 and solve the problem by replacing my macroecotools 0.4 with macroecotools 0.2 and it turns out that the new version of lognormal class make some changes of the outplow function I mentioned above. It's my mistake for taking it for granted that your paper in NEE 2017 must use the newest version of macroecotools which is upgraded in 2016. I am quite interested in applying macroecological model into microbial world and your research group has done some wonderful work concerning SADs and other ecological patterns. I am also interested in how you deal with sequence analysis before running the model 'cause all the species abundance data you collected from other projects are sequenced on different platforms and at different depth. Do you normalize or something? Otherwise, the total species abundance may vary largely and could have huge impacts on all of these models.


Best,

Jayson

**Comment by wrshoemaker** [comment]
Hi Jayson,

I'm glad that you got the code to work. This is totally on us for not including the macroecotools version in the README. Thank you for bringing it to my attention. I don't know what changes have been made to macroecotools since we published our paper, but I would recommend using the newest version for your own analyses. The Weecology group does a great job (IMO) maintaining and testing their software and I'd trust any changes that they've made.

For questions regarding sequence processing, I would ask Dr. Ken Locey, the second author on our Nature EcoEvo paper. He aggregated all the data and has a lot more experience in processing 16S sequence data for macroecological analyses than I do. For our Nature EcoEvo paper we used read counts after the error processing pipeline was completed on each study (i.e. Human Microbiome Project, Earth Microbiome Project, MGRAST). But Ken may have made changes to how he analyzes aggregated microbial sequence data since we ran these analyses. 

Total abundance does have a large impact on these models, which is why METE fails, but the lognormal succeeded. On average, increasing the size of a community results in a more uneven species abundance distribution, so models that predict a more even distribution often fail to predict the structure of microbial communities.

Best of luck! I'm looking forward to reading any microbial macroecological work you publish in the future.

Best,
Will
